File: solutions/day13.R
parse_machines <- function(machines){
    parse_matrices <- function(x){
        nums <- as.integer(x)
        list(matrix(nums[1:4], nrow = 2, byrow = FALSE), nums[5:6])
    }
    regmatches(machines, gregexpr("[0-9]+", machines))  |> 
        lapply(parse_matrices )
}

gcd <- function(a, b){
    while (b != 0){
        t  <- b
        b <- a %% b
        a <- t
    }
    a
}

solve_equation <- function(pair, cost){
    null <- c(0, 0)
    mat <- pair[[1]]
    # No rank-deficient matrices
    inverse <- matrix(c(mat[[4]], -mat[[3]], -mat[[2]], mat[[1]]), byrow  = TRUE, nrow = 2)
    lhs <- inverse %*% pair[[2]]
    determinant <- det(mat)
    lhs <- as.integer(lhs)
    determinant <- determinant
    result <- lhs %/% determinant
    print(result)
    print(ceiling(result))
    if (!identical(result, ceiling(result))) return (0)
    #browser()
  #result <- tryCatch(solve(pair[[1]], pair[[2]])  , error = function(e) null)
if(any(result > 100) || any(result < 0)){ 
    result <- null
}
crossprod(result, cost)[[1]]
}



text <- readChar("inputs/day13.txt", file.info("inputs/day13.txt")$size)  |> 
    trimws()  |> 
    strsplit("\\n\\n")  |> 
    getElement(1)

data <- parse_machines(text)
part1 <- vapply(data, solve_equation, cost = c(3, 1) , FUN.VALUE = numeric(1))  |> 
    sum()
print(part1)

File: solutions/day13.py
import re
from math import ceil
from utils.utils import split_groups, list_map


def parse(lines):
    numbers = list_map(re.findall(r"\d+", lines), int)
    return [[[numbers[0], numbers[2]], [numbers[1], numbers[3]]], numbers[4:]] 


def determinant(A):
    return (A[0][0] * A[1][1]) - (A[0][1] * A[1][0])


def check(solution, costs):
    result = 0
    for s, c in zip(solution, costs):
        if s < 0 or ceil(s) != s:
            return 0
        result += s * c
    return result

def solve_matrix(A, b):
    inverse = [[A[1][1], -A[0][1]], [-A[1][0], A[0][0]]]
    det = determinant(A)
    rhs = [
        (inverse[0][0] * b[0]) + (inverse[0][1] * b[1]),
        (inverse[1][0] * b[0]) + (inverse[1][1] * b[1]),
    ]
    return  [rhs[0] / det, rhs[1] / det]

def solve(data, prices):
    A, b = data
    first = solve_matrix(A,b)
    second = solve_matrix(A, [x + 10000000000000 for x in b])

    return [check(first, prices), check(second, prices)]


text = split_groups("inputs/day13.txt")
parsed = list_map(text, parse)
prices = [3, 1]
parts = list(zip(*[ solve(d, prices)  for d in parsed]))
part1 = int(sum(parts[0]))
part2 = int(sum(parts[1]))
print(part1)
print(part2)
